src/vidpf.rs

```diff
@@ -0,0 +1,644 @@
+
+        let CorrectionWord {
+            s_cw,
+            t_cw_l,
+            t_cw_r,
+            w_cw,
+        } = public.cw.get(level).ok_or(VidpfError::IndexLevel)?;
+
+        let seq_tilde = &self.prg(&state.seed);
+
+        let t_i = state.t_control_bit;
+        let sl = conditional_xor_seeds(&seq_tilde.0, s_cw, t_i);
+        let sr = conditional_xor_seeds(&seq_tilde.2, s_cw, t_i);
+        let tl = seq_tilde.1 ^ (t_i & *t_cw_l);
+        let tr = seq_tilde.3 ^ (t_i & *t_cw_r);
+
+        let x_i = Choice::from(u8::from(input[level]));
+        let s_tilde_i = conditional_select_seed(x_i, &[sl, sr]);
+
+        let t_new = Choice::conditional_select(&tl, &tr, x_i);
+        let (seed_new, w_i) = self.convert(&s_tilde_i);
+
+        let zero = Weight::zero(&self.weight_parameter);
+        let mut y = w_i + Weight::conditional_select(&zero, w_cw, t_new);
+        y.conditional_negate(Choice::from(key.id));
+
+        let pi_i = &state.proof;
+        let cs_i = &public.cs.get(level).ok_or(VidpfError::IndexLevel)?;
+        let pi_tilde = Self::node_proof(input, level, &seed_new)?;
+        let h2_input = pi_tilde.conditional_xor(t_new, cs_i).xor(pi_i);
+        let pi_new = Self::node_proof_adjustment(h2_input).xor(pi_i);
+
+        let state_new = EvalState {
+            level: level + 1,
+            seed: seed_new,
+            t_control_bit: t_new,
+            proof: pi_new,
+        };
+
+        Ok((state_new, y))
+    }
+
+    /// [`Vidpf::open`] recovers the input's weight from additive shares produced
+    /// by aggregation servers.
+    ///
+    /// # Errors
+    /// Triggers an error when internal proofs are not equal.
+    pub fn open(a: ValueShare<Weight>, b: ValueShare<Weight>) -> Result<Weight, VidpfError> {
+        if a.proof != b.proof {
+            return Err(VidpfError::OpenSharesFailed);
+        }
+        Ok(a.share + b.share)
+    }
+
+    fn prg(&self, seed: &Seed) -> (Seed, Choice, Seed, Choice) {
+        let mut rng =
+            XofFixedKeyAes128::seed_stream(&XofSeed(*seed), DomainSepTag::PRNG, &self.nonce);
+
+        let mut sl = Seed::default();
+        let mut sr = Seed::default();
+        rng.fill_bytes(&mut sl);
+        rng.fill_bytes(&mut sr);
+        // Use the LSB of seeds as control bits -- This ensures [prg] costs two AES calls only.
+        let tl = Choice::from(sl[0] & 0x1);
+        let tr = Choice::from(sr[0] & 0x1);
+
+        (sl, tl, sr, tr)
+    }
+
+    fn convert(&self, seed: &Seed) -> (Seed, Weight) {
+        let mut rng =
+            XofFixedKeyAes128::seed_stream(&XofSeed(*seed), DomainSepTag::CONVERT, &self.nonce);
+
+        let mut out_seed = Seed::default();
+        rng.fill_bytes(&mut out_seed);
+        let value = Weight::generate(&mut rng, &self.weight_parameter);
+
+        (out_seed, value)
+    }
+
+    fn node_proof(input: &VidpfInput, level: usize, seed: &Seed) -> Result<Proof, VidpfError> {
+        let mut binder = input.prefix(level).to_bytes();
```

Comment:
This causes an allocation, which is probably not necessary. It should be possible to iterate over the bytes of `input[..level]` and use https://docs.rs/prio/latest/prio/vdaf/xof/trait.Xof.html#tymethod.update to update the XOF state.

Reply:
this improved the timings by half.